File: Solutions/2610-closest-prime-numbers-in-range/closest-prime-numbers-in-range.js
/**
 * @param {number} left
 * @param {number} right
 * @return {number[]}
 */

var closestPrimes = function(left, right) {
    // Helper function to find all primes up to `n` using the Sieve of Eratosthenes
    const sieve = (n) => {
        const isPrime = Array(n + 1).fill(true);
        isPrime[0] = isPrime[1] = false;  // 0 and 1 are not primes
        for (let i = 2; i * i <= n; i++) {
            if (isPrime[i]) {
                for (let j = i * i; j <= n; j += i) {
                    isPrime[j] = false;
                }
            }
        }
        return isPrime;
    };

    // Get all primes up to `right`
    const isPrime = sieve(right);
    const primesInRange = [];
    
    // Collect all primes in the range [left, right]
    for (let i = left; i <= right; i++) {
        if (isPrime[i]) {
            primesInRange.push(i);
        }
    }
    
    // If less than two primes in the range, return [-1, -1]
    if (primesInRange.length < 2) {
        return [-1, -1];
    }

    // Find the pair of primes with the minimum difference
    let minDiff = Infinity;
    let result = [-1, -1];

    for (let i = 1; i < primesInRange.length; i++) {
        const diff = primesInRange[i] - primesInRange[i - 1];
        if (diff < minDiff) {
            minDiff = diff;
            result = [primesInRange[i - 1], primesInRange[i]];
        }
    }

    return result;
};

File: Solutions/2610-closest-prime-numbers-in-range/closest-prime-numbers-in-range.ex
defmodule Solution do
  @spec closest_primes(left :: integer, right :: integer) :: [integer]
  def closest_primes(left, right) do
    # Step 1: Generate all primes in the range [left..right]
    primes = Enum.filter(left..right, &is_prime?/1)

    # Step 2: If fewer than 2 primes, return [-1, -1]
    if length(primes) < 2 do
      [-1, -1]
    else
      # Step 3: Create consecutive prime pairs
      chunks = Enum.chunk_every(primes, 2, 1, :discard)

      # Step 4: Find the pair with the minimum gap
      min_pair =
        chunks
        |> Enum.min_by(fn [a, b] -> b - a end, fn -> [-1, -1] end) # fallback just in case

      min_pair
    end
  end

  # Helper function to check if a number is prime
  defp is_prime?(n) when n < 2, do: false
  defp is_prime?(2), do: true
  defp is_prime?(n) when rem(n, 2) == 0, do: false
  defp is_prime?(n) do
    max_check = :math.sqrt(n) |> floor()
    Enum.all?(3..max_check//2, fn x -> rem(n, x) != 0 end)
  end
end
